crypto-mcl.ts

```diff
@@ -0,0 +1,188 @@
+} from "./sodium";
+import * as mcl from "./mcl";
+
+export async function waitReady(){
+    await mcl.waitReady();
+    await sodiumWaitReady();
+}
+
+/**
+ * What needs to be stored in the users' phone.
+ */
+interface IUserRecord {
+    R1: mcl.G2
+    Z1: mcl.GT
+    R2: mcl.G2
+    K: Uint8Array
+    c: Uint8Array
+    nonce: Uint8Array
+}
+
+/**
+ * This is the secret information of the location.
+ */
+interface IMasterTrace{
+    mskv: mcl.Fr;
+    info: string;
+    nonce: Uint8Array;
+    mskHAEnc: Uint8Array;
+}
+
+/**
+ * Things to print on QRcodes - the mtr part only needs to be used when an infection is signaled.
+ */
+interface ILocationData {
+    // entry information
+    ent: mcl.G2;
+    // proof for the entry information
+    pEnt: Uint8Array;
+    // master tracing information for the QRtrace)
+    mtr: IMasterTrace;
+}
+
+/**
+ * The request from the health authority to the location owner to create a proof to be sent to
+ * the visitors.
+ */
+interface IPreTrace{
+    Tprime: mcl.G1
+    info: string
+    nonce: Uint8Array
+    ctxt: Uint8Array
+}
+
+/**
+ * Section7 implements the cryptographic protocol in section 7 of the CrowdNotifier whitepaper.
+ */
+export class Section7{
+    /**
+     * Setup for the HealthAuthority - returns a keypair that can be used to encrypt the QRtrack codes.
+     */
+    static setupHA(): IKeyPair{
+        return crypto_box_keypair();
+    }
+
+    /**
+     * Creates the data necessary for a location owner.
+     * @param info is the public info of the location, e.g., name:location:room
+     * @param pkH the public key of the health authority
+     * @return the LocationData to be printed in 2 (3?) QRcodes
+     */
+    static genCode(info: string, pkH: Uint8Array): ILocationData{
+        const mskv = new mcl.Fr();
+        mskv.setByCSPRNG();
+        const mskHA = new mcl.Fr();
+        mskHA.setByCSPRNG();
+        const msk = mcl.add(mskv, mskHA);
+        const mpk = mcl.mul(mcl.baseG2(), msk);
+        const nonce = randombytes_buf(msk.serialize().byteLength * 2);
+        const mskHAEnc = crypto_box_seal(mskHA.serialize(), pkH);
+        const mtr = {mskv, info, nonce, mskHAEnc}
+        return {ent: mpk, pEnt: nonce, mtr};
+    }
+
+    /**
+     * Creates a private record for the user to store in her phone.
+     * @param entry as presented in the QRentry
+     * @param proofEntry as presented in the QRentry
+     * @param info as presented in the QRentry
+     * @param counter as hours since the unix epoch
+     * @param aux free string
+     * @return user record to be stored
+     */
+    static scan(entry: mcl.G2, proofEntry: Uint8Array, info: string, counter: number, aux: string): IUserRecord{
+        const r1 = new mcl.Fr();
+        r1.setByCSPRNG();
+        const r2 = new mcl.Fr();
+        r2.setByCSPRNG();
+        const k = crypto_secretbox_keygen();
+        const nonce = secretbox_nonce();
```

Comment:
@wouterl Can we rename this ciphertext nonce to ctxt_nonce or something, just to avoid the name clash?

Reply:
done